### PR TITLE
Lecture 1 setup and product dao

### PR DESCRIPTION
**Pull request by cmakrr**

### CODE REVIEW

File: src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java [hunk 1]
```diff
@@ -1,25 +1,52 @@
...
+                .filter(product -> product.getId().equals(id))
+                .findAny()
+                .orElse(null);
     }
 
     @Override
     public List<Product> findProducts() {
-        throw new RuntimeException("Not implemented");
+        List<Product> unmodifiableProductList = productList.stream()
+                .filter(product -> product.getPrice() != null && product.getStock() > 0)
```
[the remark this answers is not in the document]
Comment (cmak33) [reply src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java]: I guess it means that product is free and this product object is considered valid. "The method shall return products having non null price and stock level > 0" (from lecture document)

File: src/main/java/com/es/phoneshop/service/ProductService.java [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+public class ProductService {
+    private final ProductDao productDao;
+
+    private ProductService() {
+        productDao = CustomProductDao.getInstance();
+    }
+
+    private static final class InstanceHolder {
+        private static final ProductService instance = new ProductService();
```
Comment (esSoftPalinaTsi) [inline src/main/java/com/es/phoneshop/service/ProductService.java:17]: same as for dao

File: src/main/java/com/es/phoneshop/service/ProductService.java [hunk 3]
```diff
@@ -0,0 +1,27 @@
+package com.es.phoneshop.service;
+
+import com.es.phoneshop.model.product.CustomProductDao;
+import com.es.phoneshop.model.product.Product;
+import com.es.phoneshop.model.product.ProductDao;
+
+import java.util.List;
+
+public class ProductService {
```
Comment (esSoftPalinaTsi) [inline src/main/java/com/es/phoneshop/service/ProductService.java:9]: service should also follow the "interface driven design"

And a note for future: we will need all the methods that are defined in dao, so please add them as well to the service, and don't forget about throwing the exceptions

Comment (esSoftPalinaTsi) [inline src/main/java/com/es/phoneshop/service/ProductService.java:9]: Unit tests need to be created for every class and method that you define

File: src/main/java/com/es/phoneshop/web/ProductListPageServlet.java [hunk 4]
```diff
@@ -1,41 +1,30 @@
...
 import jakarta.servlet.http.HttpServletResponse;
+
 import java.io.IOException;
-import java.math.BigDecimal;
-import java.util.ArrayList;
-import java.util.Currency;
-import java.util.List;
 
 public class ProductListPageServlet extends HttpServlet {
+    private ProductService productService;
```
Comment (esSoftPalinaTsi) [inline src/main/java/com/es/phoneshop/web/ProductListPageServlet.java:13]: one line after the class definition

File: src/main/java/com/es/phoneshop/service/ProductService.java [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+    private final ProductDao productDao;
```
Comment (esSoftPalinaTsi) [inline src/main/java/com/es/phoneshop/service/ProductService.java:10]: one line after the class definition

File: src/test/java/com/es/phoneshop/model/product/CustomProductDaoTest.java [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.IntStream;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+
+
+public class CustomProductDaoTest {
+    private final CustomProductDao productDao = CustomProductDao.getInstance();
```
Comment (esSoftPalinaTsi) [inline src/test/java/com/es/phoneshop/model/product/CustomProductDaoTest.java:17]: one line

File: src/test/java/com/es/phoneshop/model/product/CustomProductDaoTest.java [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+
+    private List<Product> createValidProducts(int count) {
+        List<Product> productList = new ArrayList<>();
+        BigDecimal price = new BigDecimal(1);
+        int stock = 1;
+        for (int i = 1; i <= count; i++) {
+            Product product = new Product();
```
Comment (esSoftPalinaTsi) [inline src/test/java/com/es/phoneshop/model/product/CustomProductDaoTest.java:24]: senseless creation, this creates products with same attributes
please use @Mock for Products and mock their attributes to have different values

File: src/main/java/com/es/phoneshop/model/product/ProductDao.java [hunk 8]
```diff
@@ -1,10 +1,14 @@
...
 
 public interface ProductDao {
-    Product getProduct(Long id);
+    Optional<Product> getProduct(Long id);
+
     List<Product> findProducts();
+
     void save(Product product);
+
     void delete(Long id);
```
Comment (esSoftPalinaTsi) [inline src/main/java/com/es/phoneshop/model/product/ProductDao.java:13]: any exception? What if the product with such id was no found

File: src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (esSoftPalinaTsi) [reply src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java]: Agree with you, in commerce it's possible to have free products(gift bags, wrappings) 
So if from your POV it's better to leave this ability - let's leave it as it is

File: src/main/java/com/es/phoneshop/model/product/CustomProductDao.java [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+
+    private static final class InstanceHolder {
+        private static final CustomProductDao instance = new CustomProductDao();
+    }
+
+    public static CustomProductDao getInstance() {
+        return InstanceHolder.instance;
+    }
+
+    private List<Product> createSampleProducts() {
```
Comment (esSoftPalinaTsi) [inline src/main/java/com/es/phoneshop/model/product/CustomProductDao.java:28]: please define equals and hashcode for Product

File: src/main/java/com/es/phoneshop/dao/CustomProductDao.java [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+    }
+
+    @Override
+    public void save(Product product) {
+        if (product == null) {
+            throw new NullPointerException("Product was null");
+        }
+        writeLock.lock();
+        try {
+            Optional<Product> productWithSameId = getProduct(product.getId());
```
Comment (esSoftPalinaTsi) [inline src/main/java/com/es/phoneshop/dao/CustomProductDao.java:92]: can't a deadlock happen here?
and why did you change the implementation of this method via productList.contains(product)? It was fine in while the previous review
if you want to do an update for product here, then navigating by stream is better idea than calling getProduct() inside the write lock

File: src/main/java/com/es/phoneshop/service/CustomProductService.java [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+        return instance;
+    }
+
+    @Override
+    public Product getProduct(Long id) {
+        Optional<Product> product = productDao.getProduct(id);
+        if (product.isPresent()) {
+            return product.get();
+        } else {
+            throw new NoSuchElementException(String.format("Product with id %d was not found", id));
```
Comment (esSoftPalinaTsi) [inline src/main/java/com/es/phoneshop/service/CustomProductService.java:37]: you created ProductNotFoundException, why not use it here?
